experimental/packages/opentelemetry-sdk-metrics-base/src/export/MetricReader.ts

```diff
@@ -15,51 +15,206 @@
+    timeoutHandle = setTimeout(
+      function timeoutHandler() {
+        resolve({
+          code: ReaderResultCode.TIMED_OUT,
+          error: new Error('Operation timed out.')
+        })
+      },
+      timeout
+    );
+  });
+
+  // Wrap to promise to get a result code with the result if it does not reject.
+  const resultCodePromise = promise.then(result => {
+    return { code: ReaderResultCode.SUCCESS, returnValue: result }
+  })
+
+  Promise.race([resultCodePromise, timeoutPromise]).then(result => {
+      // Clear timeout on success and return result.
+      clearTimeout(timeoutHandle);
+      if (done) {
+        done(result);
+      }
+    },
+    reason => {
+      // Clear timeout on rejection and return failure.
+      clearTimeout(timeoutHandle);
+      if (done) {
+        done({
+          code: ReaderResultCode.FAILED,
+          error: reason
+        });
+      }
+    });
+}
 
 // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader
 
+/**
+ * A registered reader of metrics that, when linked to a {@link MetricProducer}, offers global
+ * control over metrics.
+ */
 export abstract class MetricReader {
+  // Tracks the shutdown state.
   private _shutdown = false;
+  // MetricProducer used by this instance.
   private _metricProducer?: MetricProducer;
 
-  constructor(private _exporter: MetricExporter) {}
+  constructor(private readonly _preferredAggregationTemporality = AggregationTemporality.CUMULATIVE) {
+  }
 
+  /**
+   * Set the {@link MetricProducer} used by this instance.
+   *
+   * @param metricProducer
+   */
   setMetricProducer(metricProducer: MetricProducer) {
     this._metricProducer = metricProducer;
+    this.onInitialized();
   }
 
+  /**
+   * Get the {@link AggregationTemporality} preferred by this {@link MetricReader}
+   */
   getPreferredAggregationTemporality(): AggregationTemporality {
-    return this._exporter.getPreferredAggregationTemporality();
+    return this._preferredAggregationTemporality;
   }
 
-  async collect(): Promise<void> {
+  /**
+   * Handle once the SDK has initialized this {@link MetricReader}
+   *
+   * @protected
+   */
+  protected abstract onInitialized(): void;
+
+  /**
+   * Handle a shutdown signal by the SDK.
+   *
+   * <p> For push exporters, this should shut down any intervals and close any open connections.
+   * @protected
+   */
+  protected abstract onShutdown(): Promise<void>;
+
+  /**
+   * Handle a force flush signal by the SDK.
+   *
+   * <p> In all scenarios metrics should be collected via {@link collect()}.
+   * <p> For push exporters, this should collect and report metrics.
+   * @protected
+   */
+  protected abstract onForceFlush(): Promise<void>;
+
+  /**
+   * Collect all metrics from the associated {@link MetricProducer}
+   */
+  collect(options: ReaderCollectionOptions): void {
+    const timeout = options.timeoutMillis ?? 10000;
+    const done = options.done ?? (_result => {
```

Comment:
Why do we prefer callbacks to promise here?

IIUC, this might be mimicking the SpanExporter export case. However, I'm thinking that the SpanExporter is user implementable whilst the MetricReader and MeterProvider are not and we can provide a cleaner API with promise?

Reply:
Haha we had this exact discussion. He originally had an async based implementation and I advised that the APIs should work the same as the trace APIs in order to make the library feel consistent. I didn't realize that MetricReaders are not user-implementable.

Reply:
Sorry for the run-around marc :/

Reply:
If the promise pattern is preferred, we can actually reject the promise with reasons like timeout or anything else, rather than wrapping the result in a box, like ReaderResult?